core/src/consensus/vote_stake_tracker.rs

```diff
@@ -46,6 +58,60 @@ impl VoteStakeTracker {
     }
 }
 
+#[derive(Default)]
+pub struct SlotVoteTracker {
+    // Maps pubkeys that have voted for this slot
+    // to whether or not we've seen the vote on gossip.
+    // True if seen on gossip, false if only seen in replay.
+    pub voted: HashMap<Pubkey, bool>,
+    optimistic_votes_tracker: HashMap<Hash, VoteStakeTracker>,
+    pub voted_slot_updates: Option<Vec<Pubkey>>,
+    pub gossip_only_stake: u64,
+}
+
+impl SlotVoteTracker {
+    pub(crate) fn get_voted_slot_updates(&mut self) -> Option<Vec<Pubkey>> {
+        self.voted_slot_updates.take()
+    }
+
+    pub fn get_or_insert_optimistic_votes_tracker(&mut self, hash: Hash) -> &mut VoteStakeTracker {
+        self.optimistic_votes_tracker.entry(hash).or_default()
+    }
+    pub(crate) fn optimistic_votes_tracker(&self, hash: &Hash) -> Option<&VoteStakeTracker> {
+        self.optimistic_votes_tracker.get(hash)
+    }
+}
+
+pub fn initialize_progress_and_fork_choice(
```

Comment:
I don't like the location of this. This is key functionality that is run on startup and should either be part of replay stage or somewhere higher in the tvu. Doesn't make sense to associate it with the `vote_stake_tracker`